driver-core/src/main/com/mongodb/internal/logging/StructuredLogger.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.logging;
+
+import com.mongodb.connection.ClusterId;
+import com.mongodb.internal.VisibleForTesting;
+import com.mongodb.internal.diagnostics.logging.Logger;
+import com.mongodb.internal.diagnostics.logging.Loggers;
+import com.mongodb.internal.logging.StructuredLogMessage.Entry;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.mongodb.internal.VisibleForTesting.AccessModifier.PRIVATE;
+import static java.lang.String.format;
+
+/**
+ * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ */
+public class StructuredLogger {
+
+    private final Logger logger;
+
+    private static final Map<String, StructuredLoggingInterceptor> INTERCEPTORS = new HashMap<>();
+
+    public static void addInterceptor(final String applicationName, final StructuredLoggingInterceptor interceptor) {
+        INTERCEPTORS.put(applicationName, interceptor);
+    }
+
+    public static void removeInterceptor(final String applicationName) {
+        INTERCEPTORS.remove(applicationName);
+    }
+
+    public static boolean hasInterceptor(final ClusterId clusterId) {
+        return INTERCEPTORS.containsKey(clusterId.getDescription());
+    }
+
+    public static StructuredLoggingInterceptor getInterceptor(final ClusterId clusterId) {
+        return INTERCEPTORS.get(clusterId.getDescription());
+    }
+
+    public StructuredLogger(final String suffix) {
+        this(Loggers.getLogger(suffix));
+    }
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    public StructuredLogger(final Logger logger) {
+        this.logger = logger;
+    }
+
+    public boolean isDebugEnabled(final ClusterId clusterId) {
+        return logger.isDebugEnabled() || hasInterceptor(clusterId);
+    }
+
+    public void debug(final String message, final ClusterId clusterId, final String format,
```

Comment:
Many more of these messages will be added as more specs add logging.  These are just the ones need for command logging.

Reply:
Wish there was a neater / cleaner way to for this pattern.

Would `debug(final Supplier<String> logMessage, final Supplier<StructuredLogMessage> structuredLogMessage)` be cleaner?

Or even `debug(final Supplier<StructuredLogMessage> structuredLogMessage)` and have the structuredLogMessage actually handle calling logger.debug?

Reply:
I opted not to do this because there is no reason to create `StructuredLogMessage` instances at all except during testing.

Reply:
Currently, all the calls to `StructuredLogger.debug` are already guarded by `StructuredLogger.isDebugRequired`, which is why I don't understand attempts to reduce the overhead of creating objects when calling `StructuredLogger.debug`: `debug` can accept `StructuredLogMessage` as its input and it will have no performance effect as far as I can see.

Are there concerns that in the future it will become too inconvenient to do the `isDebugRequired` checks before calling `StructuredLogger.debug`? If not, then accepting `StructuredLogMessage` seems to be a better approach, and we will only need two overloads of `debug` (one that takes an exception and one that does not).

Reply:
Ah, it's because we are trying to avoid creating `StructuredLogMessage`s when debug logging is enabled but there are no interceptors (the name of `StructuredLogger.isDebugRequired` is confusing).

Reply:
The only alternative that does not have performance penalty that comes to my mind is accepting `Object... entries` (and I am not sure it's better): it can be passed to `String.format` as is, thus resulting in no overhead when debug logging is enabled but the interceptor is not present. And when the interceptor is present, we don't care about the overhead of converting it to `Entry... entries` (which is insignificant either way).

Pros: we need only two overloads of the `debug` method.
Cons: we move the checks that each name have a value and that each name is a string from compile-time to run-time.

```java
Entry[] constructEntries(final Object... deconstructedEntries) {
    assertTrue((deconstructedEntries.length & 1) == 0);
    Entry[] entries = new Entry[deconstructedEntries.length / 2];
    for (int i = 0; i < entries.length - 1; i += 2) {
        entries[i] = new Entry((String) deconstructedEntries[i], deconstructedEntries[i + 1]);
    }
    return entries;
}
```

```java
public void debug(final String message, final ClusterId clusterId, final Throwable exception, final String format,
        final Object... entries) {
    if (hasInterceptor(clusterId)) {
        getInterceptor(clusterId).intercept(new StructuredLogMessage(logger.getName(), "debug", message, clusterId, exception,
                constructEntries(entries)));
    }
    if (logger.isDebugEnabled()) {
        logger.debug(format(format, entries), exception);
    }
}
```

Reply:
I decided to go all in on `StructuredLogMessage` in the latest commit, so now there is  just a single `log` method in `StructuredLogger`.